### PR TITLE
WIP: GetRequiredPlugins for Go

### LINKED ISSUE #1549: Pulumi Go programs have no natural way of acquiring plugins
We rely on package managers to do this on install for other languages, but naturally Go can't rely on this. We should either come up with something here or (at minimum) update templates and documentation to ensure that 1) people know how to download plugins if they are missing and 2) people know to reach for `pulumi plugin install` when using a new provider with Go.

### PR DESCRIPTION
**Pull request by pgavlin**
These changes implement `GetRequiredPlugins` for Go using a registry
mechanism and an alternate entry point for `pulumi.Run`. Packages that
require plugins are expected to register themselves with the Pulumi SDK.
When `pulumi.Run` is used and the `PULUMI_PLUGINS` envvar is truthy, the
program will dump a JSON-encoded description of its required plugins to
stdout. The language host then uses this description to respond to
`GetRequiredPlugins`.

Fixes #1549.

### CODE REVIEW

**Comment by pgavlin** [comment]
I've left this as a draft b/c I'd like feedback on this approach. I think that it will work well; I've tested it out with AWS.

File: pkg/codegen/go/gen.go [hunk 1]
```diff
@@ -596,7 +596,7 @@ func (pkg *pkgContext) genResource(w io.Writer, r *schema.Resource) error {
 		fmt.Fprintf(w, "// Get%[1]s gets an existing %[1]s resource's state with the given name, ID, and optional\n", name)
 		fmt.Fprintf(w, "// state properties that are used to uniquely qualify the lookup (nil if not required).\n")
 		fmt.Fprintf(w, "func Get%s(ctx *pulumi.Context,\n", name)
-		fmt.Fprintf(w, "\tname string, id pulumi.IDInput, state *%[1]sState, opts ...pulumi.ResourceOption) (*%[1]s, error) {\n", name)
+		fmt.Fprintf(w, "\tname string, id pulumi.StringInput, state *%[1]sState, opts ...pulumi.ResourceOption) (*%[1]s, error) {\n", name)
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go]: This change is unrelated.

File: sdk/go/pulumi/run.go [hunk 2]
```diff
@@ -160,4 +173,28 @@ const (
...
+type PackageInfo struct {
+	Name    string `json:"name"`
+	Version string `json:"version,omitempty"`
+	Server  string `json:"server,omitempty"`
+}
+
+var packageRegistry = map[PackageInfo]struct{}{}
+
+func RegisterPackage(info PackageInfo) {
+	packageRegistry[info] = struct{}{}
```
Comment (lukehoban) [inline sdk/go/pulumi/run.go]: I assume it is okay for the same named package to be registered multiple times?

File: sdk/go/pulumi/run.go [hunk 3]
```diff
@@ -35,8 +37,13 @@ type RunOption func(*RunInfo)
...
-		fmt.Fprintf(os.Stderr, "error: program failed: %v\n", err)
-		os.Exit(1)
+		if err != ErrPlugins {
+			fmt.Fprintf(os.Stderr, "error: program failed: %v\n", err)
+			os.Exit(1)
+		}
+
+		printRequiredPlugins()
+		os.Exit(0)
 	}
```
Comment (lukehoban) [inline sdk/go/pulumi/run.go]: Total nit - and no change needed - but I was surprised this wasn't inverted (`if err == ErrPlugins` {`).

File: sdk/go/pulumi-language-go/main.go [hunk 4]
```diff
@@ -89,43 +156,33 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
 	if err != nil {
-		return "", errors.Wrap(err, "unable to get current working directory")
+		return nil, err
 	}
 
-	cwdProgram := filepath.Join(cwd, program)
-	if fileInfo, err := os.Stat(cwdProgram); !os.IsNotExist(err) && !fileInfo.Mode().IsDir() {
-		logging.V(5).Infof("program %s found in CWD", program)
-		return cwdProgram, nil
+	cmd.Env = []string{"PULUMI_PLUGINS=true"}
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go]: Perhaps just a comment here to make clear this is invoking the user program just to collect plugins?  This is subtle enough that worth being clear on.

**Comment by EvanBoyle** [comment]
`TestStackReferenceGo` failed in the build. I ran it locally and it passed. Going to go ahead and get this merged.
